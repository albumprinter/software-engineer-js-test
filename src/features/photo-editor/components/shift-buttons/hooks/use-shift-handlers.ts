import { useCallback, useMemo } from 'react';

import { SHIFT_STEP } from '../../../constants';
import { ShiftButtonsProps } from '../types';

type ShiftHandlers = {
  handleMoveLeft: () => void;
  handleMoveRight: () => void;
  handleMoveUp: () => void;
  handleMoveDown: () => void;
};

/**
 * Returns handlers for shift image to left, top, right or bottom.
 * @param {Object} params - Properties and function that set coordinates for shifting an image.
 * @param {HTMLImageElement|undefined} params.image - Rendered image element.
 * @param {Object} params.canvasRef - Reference to canvas element.
 * @param {Object} params.imageCoordinates - The x-axis and y-axis coordinates for placing the top-left corner of an image in the canvas.
 * @param {Function} params.setImageCoordinates - The x-axis and y-axis coordinates state setter.
 * @param {Object} params.initialCoordinates - Initial image coordinates.
 * @returns {Object} - Object with handlers for shifting an image.
 */
export const useShiftHandlers = ({
  image,
  canvasRef,
  imageSize,
  imageCoordinates,
  setImageCoordinates,
  initialCoordinates,
}: ShiftButtonsProps): ShiftHandlers => {
  const canvas = canvasRef?.current;
  const ctx = canvas?.getContext('2d');

  const { xImageAxis, yImageAxis } = imageCoordinates;
  const { imageWidth, imageHeight } = imageSize;

  const {
    current: { xImageAxis: initXImageAxis, yImageAxis: initYImageAxis },
  } = initialCoordinates;

  const imageMaxXAxisOffset = initXImageAxis * 2;
  const imageMaxYAxisOffset = initYImageAxis * 2;

  const isImageSet = image && ctx;

  /**
   * Handles moving the image to the left on a constant step per each button click.
   * @returns {void}
   */
  const handleMoveLeft = useCallback(() => {
    if (!isImageSet || xImageAxis === imageMaxXAxisOffset) {
      return;
    }

    // The constant for checking if the image is shifted to the left edge
    const shiftCoordinate = imageMaxXAxisOffset - xImageAxis + SHIFT_STEP;

    /**
     * If the image is shifted to the left edge then the offset
     * is maximal otherwise it shift on constant step
     */
    const leftXAxisCoordinate =
      shiftCoordinate >= 0 ? imageMaxXAxisOffset : xImageAxis - SHIFT_STEP;

    // Redrawing an image after a shift
    ctx.drawImage(
      image,
      leftXAxisCoordinate,
      yImageAxis,
      imageWidth,
      imageHeight
    );

    // Setting new x-axis coordinate to the local state
    setImageCoordinates((prevState) => ({
      ...prevState,
      xImageAxis: leftXAxisCoordinate,
    }));
  }, [
    ctx,
    image,
    imageHeight,
    imageMaxXAxisOffset,
    imageWidth,
    isImageSet,
    setImageCoordinates,
    xImageAxis,
    yImageAxis,
  ]);

  /**
   * Handles moving the image to the right on a constant step per each button click.
   * @returns {void}
   */
  const handleMoveRight = useCallback(() => {
    if (!isImageSet || xImageAxis === 0) {
      return;
    }

    // X-axis image coordinate after a shift
    const shiftCoordinate = xImageAxis + SHIFT_STEP;

    /**
     * If the image is shifted to the right edge then the offset
     * is equal to 0 otherwise it shift on constant step
     */
    const rightXAxisOffset = shiftCoordinate < 0 ? shiftCoordinate : 0;

    // Redrawing an image after a shift
    ctx.drawImage(image, rightXAxisOffset, yImageAxis, imageWidth, imageHeight);

    // Setting new x-axis coordinate to the local state
    setImageCoordinates((prevState) => ({
      ...prevState,
      xImageAxis: rightXAxisOffset,
    }));
  }, [
    ctx,
    image,
    imageHeight,
    imageWidth,
    isImageSet,
    setImageCoordinates,
    xImageAxis,
    yImageAxis,
  ]);

  /**
   * Handles moving the image to the down on a constant step per each button click.
   * @returns {void}
   */
  const handleMoveDown = useCallback(() => {
    if (!isImageSet || yImageAxis === imageMaxYAxisOffset) {
      return;
    }

    // The constant for checking if the image is shifted to the bottom edge
    const shiftCoordinate = imageMaxYAxisOffset - yImageAxis + SHIFT_STEP;

    /**
     * If the image is shifted to the bottom edge then the offset
     * is maximal otherwise it shift on constant step
     */
    const bottomYAxisOffset =
      shiftCoordinate >= 0 ? imageMaxYAxisOffset : yImageAxis - SHIFT_STEP;

    // Redrawing an image after a shift
    ctx.drawImage(
      image,
      xImageAxis,
      bottomYAxisOffset,
      imageWidth,
      imageHeight
    );

    // Setting new y-axis coordinate to the local state
    setImageCoordinates((prevState) => ({
      ...prevState,
      yImageAxis: bottomYAxisOffset,
    }));
  }, [
    ctx,
    image,
    imageHeight,
    imageMaxYAxisOffset,
    imageWidth,
    isImageSet,
    setImageCoordinates,
    xImageAxis,
    yImageAxis,
  ]);

  /**
   * Handles moving the image to the up on a constant step per each button click.
   * @returns {void}
   */
  const handleMoveUp = useCallback(() => {
    if (!isImageSet || yImageAxis === 0) {
      return;
    }

    // Y-axis image coordinate after a shift
    const shiftCoordinate = yImageAxis + SHIFT_STEP;

    /**
     * If the image is shifted to the top edge then the offset
     * is equal to 0 otherwise it shift on constant step
     */
    const topYAxisOffset = shiftCoordinate < 0 ? shiftCoordinate : 0;

    // Redrawing an image after a shift
    ctx.drawImage(image, xImageAxis, topYAxisOffset, imageWidth, imageHeight);

    // Setting new y-axis coordinate to the local state
    setImageCoordinates((prevState) => ({
      ...prevState,
      yImageAxis: topYAxisOffset,
    }));
  }, [
    ctx,
    image,
    imageHeight,
    imageWidth,
    isImageSet,
    setImageCoordinates,
    xImageAxis,
    yImageAxis,
  ]);

  return useMemo<ShiftHandlers>(
    () => ({
      handleMoveLeft,
      handleMoveRight,
      handleMoveUp,
      handleMoveDown,
    }),
    [handleMoveDown, handleMoveLeft, handleMoveRight, handleMoveUp]
  );
};
