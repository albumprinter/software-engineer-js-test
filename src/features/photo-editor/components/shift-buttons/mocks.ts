import { ShiftButtonsProps } from "./types";

export const mockShiftButtonsProps: ShiftButtonsProps = {
  image: new Image(),
  setImageCoordinates: jest.fn(),
  canvasRef: {
    current: document.createElement("canvas"),
  },
  imageSize: {
    imageWidth: 1440,
    imageHeight: 960,
  },
  imageCoordinates: {
    xImageAxis: -240,
    yImageAxis: -480,
  },
  initialCoordinates: {
    current: {
      xImageAxis: -280,
      yImageAxis: -280,
    },
  },
};
