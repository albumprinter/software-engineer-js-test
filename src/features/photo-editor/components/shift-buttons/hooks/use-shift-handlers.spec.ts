import { renderHook, act } from "@testing-library/react-hooks";
import "jest-canvas-mock";

import { useShiftHandlers } from "./use-shift-handlers";
import { mockShiftButtonsProps } from "../mocks";

describe("useShiftHandlers", () => {
  const ctx = mockShiftButtonsProps.canvasRef.current?.getContext("2d");

  it("should define handlers", () => {
    const {
      result: { current: handlers },
    } = renderHook(() => useShiftHandlers(mockShiftButtonsProps));

    expect(handlers).toBeDefined();
  });

  it("should shift an image to the left", () => {
    const {
      result: {
        current: { handleMoveLeft },
      },
    } = renderHook(() => useShiftHandlers(mockShiftButtonsProps));

    act(() => {
      handleMoveLeft();
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });

  it("should shift an image to the right", () => {
    const {
      result: {
        current: { handleMoveRight },
      },
    } = renderHook(() => useShiftHandlers(mockShiftButtonsProps));

    act(() => {
      handleMoveRight();
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });

  it("should shift an image to the top", () => {
    const {
      result: {
        current: { handleMoveUp },
      },
    } = renderHook(() => useShiftHandlers(mockShiftButtonsProps));

    act(() => {
      handleMoveUp();
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });

  it("should shift an image to the bottom", () => {
    const {
      result: {
        current: { handleMoveDown },
      },
    } = renderHook(() => useShiftHandlers(mockShiftButtonsProps));

    act(() => {
      handleMoveDown();
    });

    const calls = ctx?.__getDrawCalls();

    expect(calls).toMatchSnapshot();
  });
});
