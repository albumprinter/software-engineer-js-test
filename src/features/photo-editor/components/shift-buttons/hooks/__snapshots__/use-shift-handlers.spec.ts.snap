// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useShiftHandlers should shift an image to the bottom 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -265,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -215,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -240,
      "dy": -455,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -240,
      "dy": -505,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useShiftHandlers should shift an image to the left 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -265,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useShiftHandlers should shift an image to the right 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -265,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -215,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`useShiftHandlers should shift an image to the top 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -265,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -215,
      "dy": -480,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -240,
      "dy": -455,
      "img": <img />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
